src/net/server/message.rs

```diff
@@ -0,0 +1,402 @@
+///   - Finalizing the handling of a response.
+///
+/// Servers implement this trait to benefit from the common processing
+/// required while still handling aspects specific to the server themselves.
+///
+/// Processing starts at [`process_request()`].
+///
+/// [`process_request()`]: Self::process_request()
+pub trait MessageProcessor<Buf, Svc>
+where
+    Buf: BufSource + Send + Sync + 'static,
+    Buf::Output: Octets + Send + Sync + 'static,
+    Svc: Service<Buf::Output> + Send + Sync + 'static,
+{
+    type State: Clone + Send + Sync + 'static;
+
+    /// Process a DNS request message.
+    ///
+    /// This function consumes the given message buffer and processes the
+    /// contained message, if any, to completion, possibly resulting in a
+    /// response being passed to [`Self::process_call_result()`].
+    ///
+    /// The request message is a given as a seqeuence of bytes in `buf`
+    /// originating from client address `addr`.
+    ///
+    /// The [`MiddlewareChain`] and [`Service`] to be used to process the
+    /// message are supplied in the `middleware_chain` and `svc` arguments
+    /// respectively.
+    ///
+    /// Any server specific state to be used and/or updated as part of the
+    /// processing should be supplied via the `state` argument whose type is
+    /// defined by the implementing type.
+    ///
+    /// On error the result will be a [`ServiceError`].
+    fn process_request(
+        &self,
+        msg_details: MessageDetails<Buf>,
+        state: Self::State,
+        middleware_chain: MiddlewareChain<Buf::Output, Svc::Target>,
+        svc: &Svc,
+        metrics: Arc<ServerMetrics>,
+    ) -> Result<(), ServiceError>
+    where
+        Svc::Future: Send,
+    {
+        let (request, pp_res) = self.preprocess_request(
+            msg_details,
+            &middleware_chain,
+            &metrics,
+        )?;
+
+        let (txn, aborted_pp_idx) = match pp_res {
+            ControlFlow::Continue(()) => {
+                let request_for_svc = request.clone();
+
+                let res = if enabled!(Level::INFO) {
+                    let span = info_span!("svc-call",
+                        msg_id = request_for_svc.message().header().id(),
+                        client = %request_for_svc.client_addr(),
+                    );
+                    let _guard = span.enter();
+                    svc.call(request_for_svc)
+                } else {
+                    svc.call(request_for_svc)
+                };
+
+                let request_for_error = &request;
+                let txn = res.unwrap_or_else(|err| {
+                    if matches!(err, ServiceError::InternalError) {
+                        error!(
+                            "Service error while processing request: {err}"
+                        );
+                    }
+
+                    let mut response = start_reply(request_for_error);
+                    response.header_mut().set_rcode(err.rcode());
```

Comment:
There is no EDNS extended errors support in this PR.

Reply:
> There is no deliberate EDNS extended errors support in this PR. I say "deliberate" because there may be aspects of EDNS extended error support already in domain, but they are not knowingly or deliberately used by this PR.

Correction: The `BADCOOKIE` error code returned by the `CookiesMiddlewareProcessor` is an example of an extended error code, so there is some limited use of extended errors in this PR.

Reply:
You are right. Somehow I thought response was coming from somewhere else.